### PR TITLE
Feature rights and access management

### PR DESCRIPTION
**Pull request by Zukye**
UserSettings + GroupSettings

### CODE REVIEW

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java [hunk 1]
```diff
@@ -63,5 +84,33 @@ public String toString(UnitConfigType.UnitConfig object) {
         };
     }
 
+    public static void removeGroup(UnitConfigType.UnitConfig group) throws InterruptedException {
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/util/Groups.java]: How do you give the user feedback if this action fails? By this there is no possibility to do this.

File: src/main/java/org/openbase/bco/bcozy/view/InfoPane.java [hunk 2]
```diff
@@ -1,27 +1,36 @@
 /**
  * ==================================================================
- *
+ * <p>
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/view/InfoPane.java]: Please remove all generated "\<p\>" out of all license header.

File: src/main/java/org/openbase/bco/bcozy/controller/GroupController.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    private Callback<TableColumn<UnitConfigType.UnitConfig, String>, TableCell<UnitConfigType.UnitConfig, String>> cellFactory() {
+        return new Callback<TableColumn<UnitConfigType.UnitConfig, String>, TableCell<UnitConfigType.UnitConfig, String>>() {
+            @Override
+            public TableCell<UnitConfigType.UnitConfig, String> call(final TableColumn<UnitConfigType.UnitConfig, String> param) {
+                final TableCell<UnitConfigType.UnitConfig, String> cell = new TableCell<UnitConfigType.UnitConfig, String>() {
+
+                    final Button btn = new Button();
+
+                    {
+                        btn.setGraphic(new SVGIcon(FontAwesomeIcon.TIMES, Constants.EXTRA_SMALL_ICON, true));
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/controller/GroupController.java]: i have never seen such a syntax in java. Whats that for?

File: src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java [hunk 4]
```diff
@@ -55,23 +63,46 @@
...
+
+        usernameEmptyLabel.getStyleClass().remove("label");
+        firstnameEmptyLabel.getStyleClass().remove("label");
+        lastnameEmptyLabel.getStyleClass().remove("label");
+        mailEmptyLabel.getStyleClass().remove("label");
     }
 
     private void setGroups(ObservableList<UnitConfig> groups) {
         Platform.runLater(() -> usergroupField.getItems().setAll(groups));
+        System.out.println("Groups size: " + groups.size());
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java]: please cleanup

File: src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java [hunk 5]
```diff
@@ -108,32 +153,47 @@ private void register() throws InterruptedException {
             sessionManager.verifyMailAddress(mail.getText());
         } catch (VerificationFailedException ex) {
             mail.getStyleClass().add("text-field-wrong");
+            mailEmptyLabel.setVisible(true);
             return;
         }
 
         List<UnitConfig> groups = usergroupField.getCheckModel().getCheckedItems();
 
+        UserConfigType.UserConfig.Builder userConfigBuilder = UserConfigType.UserConfig.newBuilder();
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java]: Its not necessary to store this instance if never used again.

File: src/main/java/org/openbase/bco/bcozy/controller/UserSettingsController.java [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        }
+
+    }
+
+    private void initPhone() throws InterruptedException {
+        try {
+            changePhone.setText(Registries.getUserRegistry().getUserConfigById(SessionManager.getInstance().getUserId
+                    ()).getUserConfig().getMobilePhoneNumber());
+        } catch (CouldNotPerformException e) {
+            e.printStackTrace();
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/controller/UserSettingsController.java]: Please to not mess up the code by printing Stacktraces. Please use the ExceptionPrinter. Use find and replace for the hole bcozy project to replace this lines.

File: src/main/java/org/openbase/bco/bcozy/model/SessionManagerFacadeImpl.java [hunk 7]
```diff
@@ -153,5 +154,8 @@ public void verifyPhoneNumber(String phoneNumber) throws VerificationFailedExcep
     public void verifyMailAddress(String mailAddress) throws VerificationFailedException {
         //Todo check validity
         // throw new VerificationFailedException("not valid because of ...");
+        if (mailAddress.equals("")) {
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/model/SessionManagerFacadeImpl.java]: Please use isEmpty() instead

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java [hunk 8]
```diff
@@ -1,35 +1,52 @@
...
+                setGroups(Registries.getUserRegistry().getAuthorizationGroupConfigs(), groups);
+            } catch (CouldNotPerformException ex) {
+                ExceptionPrinter.printHistory(ex, LOGGER);
+            } catch (InterruptedException ex) {
+                Thread.currentThread().interrupt();
+            }
         }
 
         try {
             Registries.getUserRegistry().addDataObserver((observable, userRegistryData) ->
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/util/Groups.java]: I know this code already exists before but this is really inefficient. This observer is registered in every call of "getGroups()" but will never be removed again.

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java [hunk 9]
```diff
@@ -63,5 +84,33 @@ public String toString(UnitConfigType.UnitConfig object) {
         };
     }
 
+    public static void removeGroup(UnitConfigType.UnitConfig group) throws InterruptedException {
+        try {
+            Registries.getUserRegistry().removeAuthorizationGroupConfig(group);
+        } catch (CouldNotPerformException ex) {
+            ExceptionPrinter.printHistory(ex, LOGGER);
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/util/Groups.java]: Again, how do you inform the user if this fails?

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java [hunk 10]
```diff
@@ -63,5 +84,33 @@ public String toString(UnitConfigType.UnitConfig object) {
...
+        }
+    }
 
+    public static boolean addGroup(String groupName) throws InterruptedException {
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/util/Groups.java]: Do not use boolean return values for exception handling. throw a CouldNotPerformException instead.

File: src/main/java/org/openbase/bco/bcozy/view/CenterPane.java [hunk 11]
```diff
@@ -53,15 +51,18 @@
...
 
-    public final ObjectProperty<CenterPaneController.State> appStateProperty;
+    public ObjectProperty<CenterPaneController.State> appStateProperty;
+    private final ForegroundPane foregroundPane;
 
     /**
      * Constructor for the center pane.
      */
-    public CenterPane(final double height) {
+    public CenterPane(final double height, ForegroundPane foregroundPane) {
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/view/CenterPane.java]: final missing

File: src/main/java/org/openbase/bco/bcozy/view/InfoPane.java [hunk 12]
```diff
@@ -77,62 +91,107 @@ public Label getMouseOverText() {
...
+            infoPane.resetTimeline();
+
             Platform.runLater(() -> {
-                if (BCozy.baseColorIsWhite) {
-                    textLabel.setStyle("-fx-text-fill: white; -fx-font-size: 16;");
-                } else {
-                    textLabel.setStyle("-fx-text-fill: black; -fx-font-size: 16;");
-                }
-                textLabel.setIdentifier(identifier);
+                infoPane.setBackground(null);
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/view/InfoPane.java]: clearBackground()?

File: src/main/java/org/openbase/bco/bcozy/view/InfoPane.java [hunk 13]
```diff
@@ -77,62 +91,107 @@ public Label getMouseOverText() {
...
-            });
+            show("", "");
+            getInstance().setBackground(null);
+            getInstance().resetTimeline();
         } catch (CouldNotPerformException ex) {
             ExceptionPrinter.printHistory("Could not print user feedback!", ex, LOGGER);
         }
     }
 
-    public static void warn(final String identifier) {
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/view/InfoPane.java:110]: API brake: why did you remove this method?

File: src/main/java/org/openbase/bco/bcozy/view/InfoPane.java [hunk 14]
```diff
@@ -77,62 +91,107 @@ public Label getMouseOverText() {
...
+    public static class InfoPaneConfigurer {
 
-        } catch (CouldNotPerformException ex) {
-            ExceptionPrinter.printHistory("Could not print user feedback!", ex, LOGGER);
+        //Prevent instantiation from other classes
+        private InfoPaneConfigurer() {
         }
-    }
 
-    public static void error(final String identifier) {
```
Comment (DivineThreepwood) [inline src/main/java/org/openbase/bco/bcozy/view/InfoPane.java]: API brake: why did you remove this method?

File: src/main/resources/languages/languages_de_DE.properties [hunk 15]
```diff
@@ -4,6 +4,9 @@ startLogin=ANMELDUNG \u00d6FFNEN
 username=Benutzername
 password=Passwort
 repeatPassword=Passwort wiederholen
+changePassword=Passwort \u00e4ndern
```
Comment (DivineThreepwood) [inline src/main/resources/languages/languages_de_DE.properties]: UTF8 ?

File: src/main/resources/languages/languages_de_DE.properties [hunk 16]
```diff
@@ -4,6 +4,9 @@ startLogin=ANMELDUNG \u00d6FFNEN
...
+newPassword=Neues Passwort
```
Comment (DivineThreepwood) [inline src/main/resources/languages/languages_de_DE.properties]: Please make sure all newly introduced identifier are as well inserted into to the src/main/resources/languages/languages.properties file.

**Review by DivineThreepwood — CHANGES_REQUESTED** [review]
It is really hard to review such complex pull requests. Please work on more feature branches instead putting plenty features into one branch. For this it should be okay but further such requests will be rejected.

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java (hunk 8 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/util/Groups.java]: I dont't know how else to do this better. 
The observer is needed throughout the whole runtime of bcozy because the list of groups can be updated any time so we need to listen for changes the whole time. Don't we?

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java (hunk 1 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/util/Groups.java]: This is part of the goal I set myself for this week (as I told you last monday) and therefore not implemented in this pullrequest.

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java (hunk 9 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/util/Groups.java]: This is part of the goal I set myself for this week (as I told you last monday) and therefore not implemented in this pullrequest.

File: src/main/java/org/openbase/bco/bcozy/util/Groups.java (hunk 10 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/util/Groups.java]: this method is only used to encapsulate the thrown exception, it uses the "tryAddGroup" function that does throw the CouldNotPerformeException in case of errors.

File: src/main/java/org/openbase/bco/bcozy/view/InfoPane.java (hunk 13 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/view/InfoPane.java:110]: This method was not removed. The return type changed to InfoPaneConfigurer but all previous calls still work the same way they did before. So I would not consider this an API break?

File: src/main/java/org/openbase/bco/bcozy/view/InfoPane.java (hunk 14 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/view/InfoPane.java]: same here with the return type.

File: src/main/java/org/openbase/bco/bcozy/controller/GroupController.java (hunk 3 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/controller/GroupController.java]: It's an instance initializer 'cause you can't declare constructors in anonymous classes

File: src/main/java/org/openbase/bco/bcozy/view/InfoPane.java (hunk 12 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/view/InfoPane.java]: created clearBackground method

File: src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java (hunk 4 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java]: done

File: src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java (hunk 5 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/controller/RegistrationController.java]: fixed

File: src/main/java/org/openbase/bco/bcozy/controller/UserSettingsController.java (hunk 6 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/controller/UserSettingsController.java]: fixed

File: src/main/java/org/openbase/bco/bcozy/model/SessionManagerFacadeImpl.java (hunk 7 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/model/SessionManagerFacadeImpl.java]: done

File: src/main/java/org/openbase/bco/bcozy/view/CenterPane.java (hunk 11 above)
Comment (Zukye) [reply src/main/java/org/openbase/bco/bcozy/view/CenterPane.java]: fixed

File: src/main/resources/languages/languages_de_DE.properties (hunk 15 above)
Comment (Zukye) [reply src/main/resources/languages/languages_de_DE.properties]: As far as I know resource bundles are ISO 8859-1 encoded. Characters other than that must be declared in escaped unicode as it has been done in this project before, as well.

File: src/main/resources/languages/languages_de_DE.properties (hunk 16 above)
Comment (Zukye) [reply src/main/resources/languages/languages_de_DE.properties]: fixed

File: src/main/java/org/openbase/bco/bcozy/controller/GroupController.java (hunk 3 above)
Comment (DivineThreepwood) [reply src/main/java/org/openbase/bco/bcozy/controller/GroupController.java]: all right

**Review by DivineThreepwood — APPROVED** [review]
